### DISCUSSION TITLE
Autograding: "Resource not accesible by integration" error

### DISCUSSION THREAD

**Discussion by jzaldivar** [community/community#69948] [GitHub Education]
### Hi everyone,

Hi, everyone.

I'm having trouble with the autograding feature of GitHub Classroom.

- I have a my public template repositories for assignments in my "root" organization (jz-ayp). 
- I also have an organization for this semester's class: jz-ayp-o23.
- I created a classroom jz-ayp-o23 under the jz-apy-o23 organization.
- For that classroom (jz-ayp-o23/jz-ayp-o23), I set up an assignment using one of my public templates from my jz-ayp organization with the following options:
    - Individual assignment
    - No deadline
    - Private student repositories
    - No admin access for students to their repositories
    - Use VS Code as editor
    - No test added at assignment creation (I manually edited the classroom.yml workflow and autograding.json file) in my template and I don't want them overwritten)
- Upon accepting the assignment, students are able to clic in an "Open in VS Code" button and are prompted to install the GitHub Classroom extension.
- The problem is that, upon syncing their repository, even if all the test cases pass, they still get an "All checks failed" message because of a "Resource not accesible by integration" error they get at the end of the Autograding action
![image](https://github.com/community/community/assets/15006370/a95d0add-1f1d-4d14-914c-04011980b731)

Here is the link for the assignment: https://classroom.github.com/a/lZJcGBPs

How can I prevent that error?

Thanks in advance for your help.

**Comment by santoshvandari** [comment]
The "Resource not accessible by integration" error in GitHub Classroom typically indicates that there's a problem with the permissions or access control settings for the repositories in the assignment. To prevent this error and ensure that students can successfully complete the autograding process, you can take the following steps:

1. Check Repository Permissions:
      * Ensure that the students have appropriate permissions to access the repositories. In your assignment settings, you mentioned using "Private student repositories" and not allowing admin access for students. Double-check that students have at least "Read" access to their repositories, as "No access" or overly restrictive permissions can cause this error.
2. Verify Classroom Workflow Files:

     * Review the classroom.yml workflow file and the autograding.json file in your assignment template. Make sure that the configurations and paths are correct. Ensure that the workflow is set up to run the tests and perform autograding properly.
3. Check for GitHub Classroom Extension:

      * Ensure that students have successfully installed the GitHub Classroom extension in their VS Code. They might need to authenticate and grant necessary permissions to the extension.
4. Review Test Cases:

     * Double-check the test cases and test scripts. Ensure that they are correctly defined and that they are not causing unexpected issues. Also, confirm that the test scripts are not overly restrictive in terms of resource access.
5. Resource Access Issues:

     * Verify that your test cases and autograding scripts do not rely on resources or external services that are not accessible to the integration, or that they do not access files or directories that students do not have access to.
6. Logs and Error Messages:

     * Look at the error message in more detail to understand which resource or file is not accessible. This might provide more insight into the problem.
7. Student Communication:

     * Communicate with your students and ask them to report any issues they encounter. Encourage them to provide you with the error messages they see so that you can troubleshoot specific problems.
8. GitHub Support:

      * If you continue to face issues and are unable to resolve them, consider reaching out to GitHub Support for assistance. They can investigate the specific problem and provide guidance.
  
By carefully reviewing these aspects and ensuring that repository permissions, workflows, and test cases are correctly configured, you should be able to prevent the "Resource not accessible by integration" error in GitHub Classroom.

**Reply by jzaldivar** [reply]
Thanks for your answer.

As per your suggestions:

> 1. Check Repository Permissions.

Students have write access to their repositories.
![image](https://github.com/community/community/assets/15006370/54ceffb8-b30a-496c-b532-33511e567830)

> 2. Verify Classroom Workflow Files

Both classroom.yml and autograding.json are simple and straightforward. I don't see them using any extraneous resource. Here they are for everyone's reference:

classroom.yml:

    name: GitHub Classroom Workflow

    on:
      push:
        paths:
          - 'parte_fracc.py'
          - '.github/classroom/autograding.json'

    jobs:
      build:
        name: Autograding
        runs-on: ubuntu-latest
        steps:
          - uses: actions/checkout@v2
          - uses: education/autograding@v1
        env:
          LANG: C.UTF-8`

autograding.json:

        {
          "tests": [
            {
              "name": "Test1",
              "setup": "",
              "run": "LANG=en_US.utf8 timeout 3m python3 parte_fracc.py",
              "input": "1",
              "output": "No tiene decimales",
              "comparison": "included",
              "timeout": 1,
              "points": 1
            },
            {
              "name": "Test2",
              "setup": "",
              "run": "LANG=en_US.utf8 timeout 3m python3 parte_fracc.py",
              "input": "1.2",
              "output": "Sí tiene decimales",
              "comparison": "included",
              "timeout": 1,
              "points": 1
            },
            {
              "name": "Test3",
              "setup": "",
              "run": "LANG=en_US.utf8 timeout 3m python3 parte_fracc.py",
              "input": "-0.9",
              "output": "Sí tiene decimales",
              "comparison": "included",
              "timeout": 1,
              "points": 1
            },
            {
              "name": "Test4",
              "setup": "",
              "run": "LANG=en_US.utf8 timeout 3m python3 parte_fracc.py",
              "input": "0",
              "output": "No tiene decimales",
              "comparison": "included",
              "timeout": 1,
              "points": 1
            },
            {
              "name": "Test5",
              "setup": "",
              "run": "LANG=en_US.utf8 timeout 3m python3 parte_fracc.py",
              "input": "123.",
              "output": "No tiene decimales",
              "comparison": "included",
              "timeout": 1,
              "points": 1
            }
          ]
        }

Also, the autograding part of the action is running, and the tests are passing, as can be seen from the screen capture. It's only after the tests are run that the error occurs (as can also be seen in the screen capture):
![image](https://github.com/community/community/assets/15006370/55456006-56db-4aa5-a400-a10fc1e0bdb5)

> 3. Check for GitHub Classroom Extension

I haven't troubleshoot the students' installation but I was able to reproduce the error myself using a "fake strudent" account I setup myself using a brand-new virtual machine (AWS hosted). Same error: tests pass, resource not accessible by integration error.

> 4. Review Test Cases

As told since the beginning, test cases are passing. Check screenshoot.

> 5. Resource Access Issues

How can I check that the autograding scripts do not rely on resources or external services that are not accessible to the integration, or that they do not access files or directories that students do not have access to? I'm new to GitHub Classroom and also to GitHub, my experience is very limited in this regard.

> 6. Logs and Error Messages (Look at the error message in more detail to understand which resource or file is not accessible).

The complete error message reads: 

    Error: Resource not accessible by integration
    Error: Autograding failure: HttpError: Resource not accessible by integration

I can't, at all, get from that message which resource or file are not accessible. How can I check error logs?

> 7. Student Communication

Oh! They do communicate, and are, with reason, frustraded. They try and try to come up with a correct solution just to be greeted with an "All checs have failed!" message, even though all of the test cases are passing. I repeat: I have replicated the result using an account different than my usual.

> 8. GitHub Support:

Oh, yes, please! How can I contact GitHub support? It really is substandard and ackward to try and use a discussion forum to try and get help.

Regards,

**Comment by lwclement** [comment]
I have the same problem.  Has there been any resolution to this issue?

**Comment by markpatterson27** [comment]
`Resource not accessible by integration` is nearly always a permissions error (which I think in this case is the `checks` [^1] permission.) It is actions that needs permissions, not the student. Either a permissions section needs to be added to the workflow or workflow permissions can be changed in the org settings.

**Workflow permissions**:
Add the following to the `.github/workflows/classroom.yml` file just before the 'jobs' section.

```yaml
permissions:
  checks: write
  actions: read
  contents: read
```

For the full file, see https://github.com/markpatterson27/GitHub-Classroom-Workflows/blob/main/.github/workflows/classroom-simple.yml

OR

**Update the Org settings**:
Under `https://github.org/<your-classroom-org>/settings/actions` there is a setting for 'Workflow permissions'. This needs to be set to 'Read and write permissions'.

> ![image](https://github.com/community/community/assets/35724907/4ffbdd8a-32a6-45a2-a87b-0d7aa060e9a7)


[^1]: The `checks` permission is the one used for sending the grading points to the GH Classroom interface.
